pkg/git.go

```diff
@@ -35,16 +45,181 @@ func NewGitPackage(source *spec.GitSource) Interface {
+
+		// if subdir is provided and target is not under it, skip it
+		subDirPath := filepath.Join(prefix, subDir)
+		if subDir != "" && !strings.HasPrefix(target, subDirPath) {
+			continue
+		}
+
+		// check the file type
+		switch header.Typeflag {
+
+		// create directories as needed
+		case tar.TypeDir:
+			if _, err := os.Stat(target); err != nil {
+				if err := os.MkdirAll(target, 0755); err != nil {
+					return err
+				}
+			}
+
+		case tar.TypeReg:
+			f, err := os.OpenFile(target, os.O_CREATE|os.O_RDWR, os.FileMode(header.Mode))
+			if err != nil {
+				return err
+			}
+
+			// copy over contents
+			if _, err := io.Copy(f, tr); err != nil {
+				return err
+			}
+			f.Close()
+		}
+	}
+}
+
 func (p *GitPackage) Install(ctx context.Context, dir, version string) (lockVersion string, err error) {
-	cmd := exec.CommandContext(ctx, "git", "clone", p.Source.Remote, dir)
+	// Optimization for GitHub sources: download a tarball archive of the requested
+	// version instead of cloning the entire repository. Resolves the version to a
+	// commit SHA using the GitHub API.
+	if strings.HasPrefix(p.Source.Remote, "https://github.com/") {
+		archiveUrl := fmt.Sprintf("%s/archive/%s.tar.gz", p.Source.Remote, version)
+		archiveFilepath := fmt.Sprintf("%s.tar.gz", dir)
+
+		defer os.Remove(archiveFilepath)
+		commitSha, err := downloadGitHubArchive(archiveFilepath, archiveUrl)
+		if err != nil {
+			return "", err
+		}
+		r, err := os.Open(archiveFilepath)
+		err = gzipUntar(dir, r, p.Source.Subdir)
+		return commitSha, nil
+	}
+
+	cmd := exec.CommandContext(ctx, "git", "init")
 	cmd.Stdin = os.Stdin
 	cmd.Stdout = os.Stdout
 	cmd.Stderr = os.Stderr
+	cmd.Dir = dir
 	err = cmd.Run()
 	if err != nil {
 		return "", err
 	}
 
+	cmd = exec.CommandContext(ctx, "git", "remote", "add", "origin", p.Source.Remote)
+	cmd.Stdin = os.Stdin
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	cmd.Dir = dir
+	err = cmd.Run()
+	if err != nil {
+		return "", err
+	}
+
+	// Attempt shallow fetch at specific revision
+	cmd = exec.CommandContext(ctx, "git", "fetch", "--depth", "1", "origin", version)
+	cmd.Stdin = os.Stdin
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	cmd.Dir = dir
+	err = cmd.Run()
+	if err != nil {
+		// Fall back to normal fetch (all revisions)
+		cmd = exec.CommandContext(ctx, "git", "fetch", "origin")
+		cmd.Stdin = os.Stdin
+		cmd.Stdout = os.Stdout
+		cmd.Stderr = os.Stderr
+		cmd.Dir = dir
+		err = cmd.Run()
+		if err != nil {
+			return "", err
+		}
+	}
+
+	// Sparse checkout optimization: if a Subdir is specificied,
+	// there is no need to do a full checkout
+	if p.Source.Subdir != "" {
+		cmd = exec.CommandContext(ctx, "git", "config", "core.sparsecheckout", "true")
+		cmd.Stdin = os.Stdin
+		cmd.Stdout = os.Stdout
+		cmd.Stderr = os.Stderr
+		cmd.Dir = dir
+		err = cmd.Run()
+		if err != nil {
+			return "", err
+		}
+		glob := []byte(p.Source.Subdir + "/*\n")
+		ioutil.WriteFile(dir+"/.git/info/sparse-checkout", glob, 0644)
```

Comment:
It seems unlikely that writing a file inside the dir would fail at this point with the previous operations succeeding. In any case, the entire repository is a tmpDir that will be wiped out on every install/update, so I think in this case the best option is to bail out with the error.